Migrate parts of experimental API into APIv3

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/18491
## Description

This PR aims at providing everything that is needed to render the [WP list](https://community.openproject.org/projects/openproject/work_packages) in APIv3. The WP list is currently mainly built from the experimental API, a very specific API just intended for consumption by the WP list.

APIv3 has a broader audience (it should be a general purpose API), but is not yet able to serve all functions required to render the WP list.

Functionality provided in experimental API (and _apparently_ consumed in WP list) includes:
- [x] work package sorting (collection filters/options)
- [x] work package filtering (collection filters)
- [x] work package grouping (collection aggregation)
  - [x] `valueLink`
- [x] building sums (collection aggregation)
  - [x] TODO: correct sums for durations
- [x] TODO: move sums and groups out of `_embedded` zone
- [x] TODO: validate user inputs above
  - [x] error handling
- [x] non project-scoped work package access
- [ ] column meta information (schema)
  - [x] endpoint for project-wise WP schema (free of type information)
  - [x] exclude `lockVersion` column
  - [ ] ~~include groupability information~~
- [x] values of settings (configuration)
  - [x] page-size steps aka `per_page_options`
- [ ] ~~available export formats, e.g. Feeds, PDF, etc (???)~~
- [x] TODO: convert field names (see caveats below)
- [x] TODO: adapt plugins (e.g. WP schema)

Where possible I entered existing concepts of APIv3 in parenthesis, that may be able to fullfill the purpose (by being extended)...

I still need to check which of those things are actually needed by the client, because there seems to be a lot of redundancy...
APIv3 does also not yet have concepts for all of the functionality provided (e.g. aggregation)...
## Caveats

For the parameters `groupBy`, `sortBy` and `filters` I decided to accept the same formats as returned by the query representer (this would also be my goal with a QL, just that the representer would return the QL, rather than a custom JSON format). ~~However, since the query representer returns colum names as they are refered to internally to OpenProject, this does not always match up with the property names you would expect from the APIv3. Examples:~~
- ~~`assignee` vs `assigned_to`~~
- ~~`estimatedTime` vs `estimated_hours`~~

~~I really would like to accept "API names" here, but I fear this is out of scope of this PR. I therefore decided to accept internally used active record names rather than "API names" for those parameters, because this is whats stored and returned in our queries and also whats consumed by the downstream code, used to actually perform the filtering/grouping.~~ edit: I actually included the name conversion in this PR, as inconsistent names would probably have lead to an unusable API, because the client will also seed the names of new columns etc from the API.

All other results from the API (including sums etc) should use the property names known from the API.
## relevant Plugin-PRs
- https://github.com/finnlabs/openproject-backlogs/issues/158
- https://github.com/finnlabs/openproject-costs/issues/160

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-migrate-wp-endpoint-from-experimental.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-migrate-wp-endpoint-from-experimental.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
offer setting for page sizes via config endpoint

NobodysNightmare committed:
adapt documentation to truth w.r.t. filters & pagination

- available_* endpoints are not paginatable
- WP list is paginatable
- there are no filters supported (and the ones coming soon are not OQL)

NobodysNightmare committed:
allow filtering using JSON from query representer

NobodysNightmare committed:
allow to get queries via APIv3

- one of those missing endpoints
- adapted resource spec to be closer to other resource specs
- specified error behaviour (missing permissions)
-> change: not being allowed to see a query results in 404

NobodysNightmare committed:
basic specification of filters for WP list

NobodysNightmare committed:
pass block parameter through :authorize

NobodysNightmare committed:
allow sorting of WP lists

NobodysNightmare committed:
extract WP collection methods into helper class

NobodysNightmare committed:
add global WP list endpoint

NobodysNightmare committed:
correctly camelize query parameter

NobodysNightmare committed:
add specification for grouping and sums in APIv3

NobodysNightmare committed:
remove superfluous length call

NobodysNightmare committed:
allow to calculate property sums

NobodysNightmare committed:
some work towards readability

NobodysNightmare committed:
always include relevant columns

- caller does not need to do it anymore
- caller is still allowed to do it, because it can also add own conditions...

NobodysNightmare committed:
implement grouping

NobodysNightmare committed:
some code structuring

NobodysNightmare committed:
fix and extend collection representer specs

NobodysNightmare committed:
add some integrative specs for sum/group/sort/filter

NobodysNightmare committed:
improve error handling for JSON parsing parameters

NobodysNightmare committed:
update documentation w.r.t. showSums and groupBy

NobodysNightmare committed:
improve rubocop-perceived styling

NobodysNightmare committed:
include link to linkable group values

NobodysNightmare committed:
extract groups into own representer

NobodysNightmare committed:
fix reference to value links

NobodysNightmare committed:
don't allow display_sums to be NULL

NobodysNightmare committed:
align the query representer with APIv3

- correctly use datatypes
- link to linkable resources (instead of embedding their properties)
- add specs

NobodysNightmare committed:
convert attribute names in queries API

This should make the API more consistent again, accepting and producing the same attribute names on all endpoints

NobodysNightmare committed:
move writability decision into WP schema

- reducing the method count (aka interface size) of the schema
- giving it more control
- still allow outside override in representer (for plugin compatibility)

NobodysNightmare committed:
reformat SQL

NobodysNightmare committed:
ensure sortability of all user columns

I am not seeing any other unsortable columns, so this might make the API exposed "sortable" property obsolete (exposed in experimental API)

NobodysNightmare committed:
fix specs

- also extend WP schema specs

NobodysNightmare committed:
fix expectations in legacy spec

- real code is actually able to deal with arrays being returned from sortable
- simplified group by sorting

NobodysNightmare committed:
split implementations for work package schema

- was already quite case-ish, depending on its creation via WP or via project + type
- I plan on adding a third case, which would not improve the situation for a single class

NobodysNightmare committed:
add endpoint to list work package columns in a project

oliverguenther commented on app/controllers/api/experimental/concerns/column_data.rb:
```diff
@@ -180,6 +180,7 @@ def sum_columns(column_names, work_packages)
     end
   end
 
+  # FIXME: Deprecated. Should be removable once the experimental WP controller is gone
```
So was the comment simply wrong are are you going to remove the method later in this PR? :wink:

NobodysNightmare replied:
Oh, that is the sad part of this PR... It will only add stuff, but not remove the experimental API, because it is still being used...

oliverguenther commented on app/models/query/sums.rb:
```diff
@@ -76,7 +76,7 @@ def total_sum_of(column)
 
   def sum_of(column, collection)
-    return unless should_be_summed_up?(column)
+    return nil unless should_be_summed_up?(column)
```
I'm all for explicit, but these two are identical, aren't they?

NobodysNightmare replied:
They are... I guess I would usually not have changed it (because I touched nothing else in that method).
But my reasoning is something like: If a method usually returns a non-nil value, be explicit about it returning nil.

oliverguenther commented on doc/apiv3-documentation.apib:
```diff
@@ -1140,6 +1215,7 @@ For now this endpoint will only allow access to settings deemed useful for a cli
 | Property                  | Description                                        | Type       | Condition                         | Supported operations |
 | :-----------------------: | -------------------------------------------------- | ---------- | --------------------------------- | -------------------- |
 | maximumAttachmentFileSize | The maximum allowed size of an attachment in Bytes | Integer    |                                   | READ                 |
+| perPageOptions            | Pag size steps to be offered in paginated list UI  | Integer[]  |                                   | READ                 |
```
As I commented on an earlier commit before,again here: `Pag size` typo

NobodysNightmare committed:
typo

[ci skip]

NobodysNightmare committed:
replace deprecated AR method

NobodysNightmare committed:
fix stupid C&P error

NobodysNightmare committed:
improve code style

NobodysNightmare committed:
hide lockVersion from work package columns

oliverguenther commented on lib/api/v3/queries/queries_api.rb:
```diff
@@ -41,14 +41,27 @@ class QueriesAPI < ::API::OpenProjectAPI
             before do
               @query = Query.find(params[:id])
               @representer = QueryRepresenter.new(@query)
+              authorize_as_not_found(:show)
```
Was wondering what that would actually mean, I don't find the naming to be precise. Not saying that I could come up with the best suggestion, but was thinking about `authorize_or_not_found`.

NobodysNightmare replied:
You are right... I slightly modified `authorize_by_policy`, it will now behave like `authorize` methods from `root.rb` and accept a block:

``` ruby
authorize_by_policy(:show) do
  raise API::Errors::NotFound
end
```

oliverguenther commented on lib/api/v3/work_packages/work_package_list_helpers.rb:
```diff
@@ -0,0 +1,194 @@
+        #   },
+        #   { /* more filters if needed */}
+        # ]
+        def set_filters_from_json(query, json)
```
You should extract parsing and applying of sort and filters into a separate helper, which you pass the query object. I don't think it belongs in this helper.

oliverguenther commented on app/models/custom_field.rb:
```diff
@@ -187,25 +187,40 @@ def cast_value(value)
   # objects by their value of the custom field.
   # Returns false, if the custom field can not be used for sorting.
-  def order_statement
+  def order_statements
```
Certainly not your fault as that's how the custom fields work, but there's a lot of redundancy in the queries you changed below. They actually share a lot of common knowledge. It may not be practical, but what do you think about building these queries in a helper and adding the custom constraints for specific types to the otherwise identical query?

oliverguenther commented:
I'm actually pleasantly surprised at how well one could go through the changes in the PR, but not too surprised given that you are quite at home at the API @NobodysNightmare. The majority applies quite well to the existing APIv3 functionality and I'm glad you went the extra mile to consolidate all field names in the AR / API conversion helper.

There are a few issues regarding complexity, but nothing extraordinary for the size of the changes introduced here.

Since my contact with the APIv2 has been very limited to non-existing, I feel I'm not entitled to commenting on questions of completeness and compatibility, especially regarding plugins. But, If I recall your comments correctly, APIv2 will remain for the time being, so this isn't the particular focus of this PR.

NobodysNightmare committed:
refactor 404 auth error

no separate method needed, let authorize_by_policy behave like well known auth methods (accepting a block)

NobodysNightmare committed:
split parsing and applying of filters/sort

NobodysNightmare committed:
extract order_statements out of CustomField

NobodysNightmare committed:
fix spec double and remove users from groupable CFs

- spec double was not up to date with reality of return values
- user custom fields are not successfully groupable, because the order statement is a list

NobodysNightmare committed:
remove groupability from docs again

there are probably better solutions to be evaluated (e.g. a Form for queries)

NobodysNightmare commented:
@oliverguenther I think to have addressed your concerns, please check whether I did things in a way you like :)

General note: I'd like to finish this PR within its current scope. This will allow for implementation of the largest parts of the work package list using APIv3.

Omissions that I am aware of:
- **export formats**: I did not yet define a way to let the client discover supported export formats for the work package list. This is a functionality provided by the experimental API and not yet covered by APIv3.
- **groupable columns**: While a client can group the result by any column it desires, the client can't find out which columns support grouping. I initially anticipated to put this specific information into the work package schema, giving a `groupable` attribute to each property. I talked about his with @ulferts, because this would have caused a link between `Query` and the schema. His idea was interesting enough: Allow to create/edit queries via API and provide a form that aids in their creation. This form could tell me which values are valid for `groupBy`, just like I know it from the WP form.

Whatever solution we choose for those problems, this PR already provides a great enhancement over the status quo and we can implement the rest in a separate PR (IMHO).

That is why I am opening this PR for final merge-review (I will probably have to talk to travis first, thogh).

NobodysNightmare committed:
reintroduce form_embedded argument for schema

we can't live without it :/

NobodysNightmare committed:
move auth check out of before block

different endpoints under work_packages require different permissions

NobodysNightmare commented:
Just to make this does not get lost:

**yes** review this PR

~~**no** don't merge it, because I will have to prepare some plugins to stay compatible with the changes made here.~~ Plugins are looking fine. Review and merge me!

NobodysNightmare committed:
adapt acts as op engine to define methods via def

- putting instance methods into InstanceMethods module is consistent with other ActsAs* implementations
- the InstanceMethodsModule also allows me to reduce the diff size, because I keep the indentation of before

NobodysNightmare committed:
change attribute registration via plugin to be more generic

- now also registers the attribtue as writable in the schema

NobodysNightmare committed:
simplify interface for assignable values

- one method to rule them all
- eases overriding
  -> does not require to define method for new types on base class
- allows to infer values_callback from property name (schema representer)

oliverguenther commented:
> @oliverguenther I think to have addressed your concerns, please check whether I did things in a way you like :)

Looks _very_ good. Thanks for the extra mile. I give this :+1: code wise and move on to giving this one and the affected plugins a whirl.

oliverguenther commented:
Something not directly tied to your PR: With the large number of parameters, the documentation style of a API endpoint overflows:

![bildschirmfoto 2015-08-27 um 09 30 55](https://cloud.githubusercontent.com/assets/459462/9514969/86b9cb86-4c9e-11e5-991a-52faf1618375.png)

A simple solution would be to change this to `/api/v3/work_packages{?parameters}` or something.

NobodysNightmare commented:
That is unfortunate...

> A simple solution would be to change this to `/api/v3/work_packages{?parameters}` or something.

Not really, because API blueprint forces me to specify the possible parameters that way (at least IIRC). This is because nearly everything is semantically parsable and therefore specified. It is only the free text sections in between where I can write what I want ^^

One could argue that the renderer (aglio) should behave better when displaying it... For now I'd like to leave it that way. Possible workarounds include:
- update aglio and see what happens
- provide a different CSS style for the generated documentation, that does not suffer from that weakness

oliverguenther commented:
Ah, I see. Yes, I believe that is a matter for another day. I'll keep a note and create a ticket for the wish list. It isn't exactly a catastrophe, either.

Another question entirely. When providing invalid arugments, such as the following filter `http://localhost:5000/api/v3/work_packages?filters=[{ "type":  { "operator": "=", "values": ["Type"]} }]`, 
I'm getting a raw AR exception (and thus a 500 internal on production).

```
lib/api/decorators/offset_paginated_collection.rb:50:in `initialize'
lib/api/v3/work_packages/work_package_collection_representer.rb:52:in `initialize'
lib/api/v3/work_packages/work_package_list_helpers.rb:177:in `new'
lib/api/v3/work_packages/work_package_list_helpers.rb:177:in `collection_representer'
lib/api/v3/work_packages/work_package_list_helpers.rb:49:in `work_packages_by_params'
lib/api/v3/work_packages/work_packages_api.rb:40:in `block (2 levels) in <class:WorkPackagesAPI>'
app/middleware/params_parser_with_exclusion.rb:38:in `call'
```

We could probably try to catch these errors high up in the API and return a meaningful query error, as happens when e.g., the JSON itself is invalid.

What do you think? Is this too complex? 

---

Same goes when using something like `...?filters=[{ "type":  { "operator": "=", "values": [0]} }]`, causing an ArgumentError deep within the Query model.

```
app/models/query.rb:509:in `block in sql_for_field'
app/models/query.rb:509:in `map'
app/models/query.rb:509:in `sql_for_field'
app/models/query.rb:451:in `block in statement'
app/models/query.rb:359:in `each'
app/models/query.rb:359:in `statement'
app/models/query/results.rb:81:in `work_packages'
app/models/query/results.rb:94:in `sorted_work_packages'
lib/api/v3/work_packages/work_package_list_helpers.rb:49:in `work_packages_by_params'
lib/api/v3/work_packages/work_packages_api.rb:40:in `block (2 levels) in <class:WorkPackagesAPI>'
app/middleware/params_parser_with_exclusion.rb:38:in `call'
```

NobodysNightmare commented:
Good point. When it came to error handling I was not very creative in how I could break it.

I hate HTTP 500 appearing in such cases, especially since the way our filters work is already asking for trouble. I will see how far I can get with some more rescue etc. But some errors are completely hidden by the underlying code (e.g. `filters=[{"status": { ... }, "type": { ... }}]` looks like a perfectly valid filter list, but IIRC will be silently ignored, because there are some rather harsh type checks involved.

NobodysNightmare commented:
@oliverguenther For your first example (`http://localhost:5000/api/v3/work_packages?filters=[{ "type": { "operator": "=", "values": ["Type"]} }]`) I am getting an empty collection as response, no errors...

NobodysNightmare committed:
rename parameter/update code doc

it is actually expected to be an array...

oliverguenther commented:
Hmm.. (pg) I'm getting `PG::InvalidTextRepresentation: ERROR: invalid input syntax for integer: "Type" LINE 1: ...` because it explicity checks the `type_id` with type: `... WHERE (((work_packages.type_id IN ('Type')) ...`

oliverguenther commented:
There's also things caused by my own naïve experiments, such as using `?groupBy="status"` (shouldn't have quotes, that was my fault reading the documentation wrong, but still), which is parsed as `"\"status\""`, or `?groupBy=thisdoesntexist`, which both throw:

```
undefined method `map' for nil:NilClass
lib/api/v3/work_packages/work_package_list_helpers.rb:126:in `generate_groups'
lib/api/v3/work_packages/work_package_list_helpers.rb:120:in `apply_and_generate_groups'
lib/api/v3/work_packages/work_package_list_helpers.rb:42:in `work_packages_by_params'
lib/api/v3/work_packages/work_packages_api.rb:40:in `block (2 levels) in <class:WorkPackagesAPI>'
app/middleware/params_parser_with_exclusion.rb:38:in `call'
```

oliverguenther commented:
Using `?groupBy=status` takes roughly 10s to execute that query for my 63 work packages (with a collection page size of 30), probably due to the N Journal::WorkPackageJournals even when cached.

NobodysNightmare committed:
increase type compatibility of :sql_for_field

it is expecting a string array anyway (and casting to integer on demand)
still that makes ints not less valid...

NobodysNightmare commented:
I also saw those massive journal queries lately (without any grouping)...

I should probably investigate that (not sure if or if not PR related)

oliverguenther commented:
I would agree to leaving it out of this PR.

There seems to be some caching issue. I've uploaded a gist with a full log for one query (limited to 10 items)

https://gist.github.com/oliverguenther/9090251baee2c48f0438

I've tested filters, groupBy and sortBy on both the global and per-project collection, that seems to work fine except for the aforementioned exception issues.

I'm having difficulties in understanding aggregation queries. That is returned when combining `groupBy` and `showSums` and depends on whether the column of groupBy can be summarized?

I can't seem to produce a query where `totalSums` isn't empty..

NobodysNightmare commented:
Okay, I could reproduce the error with your first query using PSQL. This error is quite sad:

The offending filter is built in the shared helper, but the exception occurs when actually executing the query (in the offset paginated collection), so I am really out of reach to detect that error using a `rescue`.

I also can't test for stringyness/intiness while parsing the filter, because for some attribtues strings are okay, for others not (subject vs type). Still I don't find it acceptable to raise an HTTP 500, even though it is the best thing to do from the perspective of the `OffsetPaginatedCollection`...

I'll probably have to think about that a little more.

At least I could find a fix for your second query ^^

> I can't seem to produce a query where `totalSums` isn't empty..

haha, I made the same mistake ^^ Check out http://localhost:5000/settings?tab=work_packages and add some summable fields there. However, I wasn't brave enough to try making a version summable ^^

NobodysNightmare committed:
clarify documentation on summable columns

NobodysNightmare commented:
> There seems to be some caching issue.

Urg of course!
The problem is that we include all the journals for a work package in its representer. When you request 30 work packages, for each of those work packages, it will get the aggregated journals via SQL.

Fortunately this is a problem present outside of this PR. Unfortunately we will have to care about it nevertheless.

My take is still to leave embedded properties of any kind out of collection responses and only return links...

Azure7111 commented on app/models/query.rb:
```diff
@@ -475,24 +496,27 @@ def custom_field_id(filter)
+  # Helper method to generate the WHERE sql for a +field+, +operator+ and a +values+ array
+  def sql_for_field(field, operator, values, db_table, db_field, is_custom_filter = false)
+    # code expects strings (e.g. for quoting), but ints would work as well: unify them here
+    values = values.map { |value| value.to_s }
```
You might want to give this one a spin.

NobodysNightmare commented on app/models/query.rb:
```diff
@@ -475,24 +496,27 @@ def custom_field_id(filter)
-  # Helper method to generate the WHERE sql for a +field+, +operator+ and a +value+
-  def sql_for_field(field, operator, value, db_table, db_field, is_custom_filter = false)
+  # Helper method to generate the WHERE sql for a +field+, +operator+ and a +values+ array
+  def sql_for_field(field, operator, values, db_table, db_field, is_custom_filter = false)
```
It's over 9000!!!!

oliverguenther commented:
> My take is still to leave embedded properties of any kind out of collection responses and only return links...

:+1:

oliverguenther commented:
Just to clear things up here:

`http://localhost:5000/api/v3/work_packages?filters=[{ "type": { "operator": "=", "values": ["Type"]} }]` still fails for me, but your fix allows me to write either:

`http://localhost:5000/api/v3/work_packages?filters=[{ "type": { "operator": "=", "values": [1]} }]`
or
`http://localhost:5000/api/v3/work_packages?filters=[{ "type": { "operator": "=", "values": ["1"]} }]`

NobodysNightmare commented:
exactly!

oliverguenther commented:
Well, all is dandy then.

We have covered:
- [x] the caching issue of journals (out of scope. will be addressed in separate PR)
- [x] removing embedded properties in (WP) collections. Also moved to a separate PR.
- [x] Exception handling of wrong filters etc. I agree that this can only be addressed when issues come up but I'm still a little undecided on that end. Users should not receive a 500 through the API whenever possible.

NobodysNightmare committed:
in production handle internal API errors gracefully

- still allow for useful errors during development
- let all own exceptions inherit from StandardError, the thing you should use for own errors...

NobodysNightmare commented:
@oliverguenther As discussed I've added a catch-all exception handler for production mode. This should at least make internal errors "nice" for all real systems.

During development we stay with our default Rails error pages, this allows to still get stacktraces and other useful information when you messed up.

oliverguenther commented:
Cool. Will give this one last whirl and will merge when travis agrees.
